Pro 5369 solicitors as executor in grant flow

jackmunkonge opened the pull request:
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRO-5369


### Change description ###
Solicitors as executor in grant flow


**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

sanjayparekh-hmcts commented on src/main/java/uk/gov/hmcts/probate/controller/BusinessValidationController.java:
```diff
@@ -66,6 +66,11 @@
 
+    @PostMapping(path = "/sols-apply-as-exec")
+    public ResponseEntity<CallbackResponse> solsApplyAsExec(@RequestBody CallbackRequest request) {
+        return ResponseEntity.ok(callbackResponseTransformer.toggleApplicantFields(request));
```
rename to setApplicantFieldsForSolsApplyAsExec

sanjayparekh-hmcts commented on src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java:
```diff
@@ -552,9 +571,93 @@ private boolean isPrimaryApplicantNotApplying() {
+        return (solsWillType != null);
+    }
+
+    private List<CollectionMember<AdditionalExecutor>> mapAdditionalExecutorsApplying(List<CollectionMember<AdditionalExecutorApplying>> additionalExecutors) {
```
please make sure additional functionality is covered by unit tests for this class

sanjayparekh-hmcts commented on src/main/java/uk/gov/hmcts/probate/model/ccd/raw/request/CaseData.java:
```diff
@@ -552,9 +571,93 @@ private boolean isPrimaryApplicantNotApplying() {
+
+            if (exec.getApplyingExecutorFirstName() == null || exec.getApplyingExecutorLastName() == null) {
+                List<String> names = splitFullname(exec.getApplyingExecutorName());
+                surname = names.remove(names.size()-1);
```
Check double barrelled sur names and hyphenated fore/surnames work as expected. Also single names (eg. Madonna)

sanjayparekh-hmcts commented on src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java:
```diff
@@ -581,6 +593,10 @@ private boolean isCodicil(CaseData caseData) {
         return YES.equals(caseData.getWillHasCodicils());
     }
 
+    private boolean isApplyAsExecutor(CaseData caseData) {
```
isSolicitorApplyingAsExecutor

sanjayparekh-hmcts commented on src/main/java/uk/gov/hmcts/probate/transformer/CallbackResponseTransformer.java:
```diff
@@ -831,6 +877,24 @@ private void updateCaseBuilderForTransformCase(CaseData caseData, ResponseCaseDa
         }
 
+        if (isApplyAsExecutor(caseData)) {
+            if (YES.equals(caseData.getSolsSolicitorIsMainApplicant())) {
```
see if you can refactor this function into its own method for both calls

jackmunkonge replied:
Ends up making code less readable; having to pass multiple mutable objects as arguments

sanjayparekh-hmcts commented on src/test/java/uk/gov/hmcts/probate/controller/BusinessValidationControllerTest.java:
```diff
@@ -190,6 +192,10 @@ public void setup() {
                 .solsSolicitorAddress(solsAddress)
                 .ihtFormId(IHT_FORM)
                 .solsSOTName(SOLICITOR_NAME)
+                .solsSolicitorIsApplyingExec(YES)
```
check tests for these values being null or NO

jackmunkonge committed:
PRO-5370: Add sols apply as exec

PRO-5371: Add solsSolicitorApplying to definitions


PRO-5371: Add to amend case details


PRO-5371: Amend case details full refactor


PRO-5370: Fixes


PRO-5371: Amend case details upate


PRO-5370: Add isMainApplicant and tests update


PRO-5371: Update event to case fields


PRO-5372: Change to solsSolicitorIsApplying


PRO-5372: Move sol name to after question


PRO-5372: Add SOT forenames and surname


PRO-5372: Add sols names to code and join to solsSOTName


PRO-5372: Conditionally display applicant details


PRO-5372: Revert SoT name for amend details


PRO-5372: Add isApplying and notApplyingReason


wefwefw


ewfef


wefww


PRO-5372: Readd sot forenames and surname to amend events


nbhbjb


PRO-5372: Don't display primaryapplicant fields for sols primary


segre


wefwef


PRO-5372: Don't save pdf write


PRO-5372: Fix callbackresponsetransformer for sols


PRO-5372: Update sols firm details


PRO-5372: Refactored applicant fields toggling


PRO-5372: ferfe


PRO-5372: Authorise solicitorIsMainApplicant


khjghc


PRO-5372: Fix unit tests


PRO-5372: fdsdf


PRO-5372: Fix unit tests


PRO-5372: Always show applying question when sols applying as exec


PRO-5372: Show solsapplying and notapplyingreason in tabs


PRO-5372: Stop SOT name from nulling


PRO-5372: Bug fixes


PRO-5369: Fix rebase


PRO-5372: Add units

jackmunkonge committed:
PRO-5369: Fix change requests


egrere
